KSystemInformer/object.c

```diff
@@ -2372,6 +2372,41 @@ NTSTATUS KphOpenNamedObject(
     return status;
 }
 
+/**
+ * \brief Opens a named object by its type index from nt!ObTypeIndexTable.
+ *
+ * \param[out] ObjectHandle Set to the opened handle.
+ * \param[in] DesiredAccess Desires access to the object.
+ * \param[in] ObjectAttributes Attributes to open the object.
+ * \param[in] TypeIndex Type index of object to open.
+ * \param[in] AccessMode The mode in which to perform access checks.
+ *
+ * \return Successful or errant status.
+ */
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Must_inspect_result_
+NTSTATUS KphOpenObjectByTypeIndex(
+    _Out_ PHANDLE ObjectHandle,
+    _In_ ACCESS_MASK DesiredAccess,
+    _In_ POBJECT_ATTRIBUTES ObjectAttributes,
+    _In_ ULONG TypeIndex,
+    _In_ KPROCESSOR_MODE AccessMode
+)
+{
+    KPH_PAGED_CODE_PASSIVE();
+
+    if (!KphDynObTypeIndexTable)
+    {
+        return STATUS_NOINTERFACE;
+    }
+
+    return KphOpenNamedObject(ObjectHandle,
+                              DesiredAccess,
+                              ObjectAttributes,
+                              KphDynObTypeIndexTable[TypeIndex],
```

Comment:
I thought that it enough to validate parameters in native.c
Is it correct to check this way from the kernel?
```
    if (AccessMode != KernelMode)
    {
        OBJECT_TYPES_INFORMATION typesInfo = { 0 };

        if (ZwQueryObject(NULL,
            ObjectTypesInformation,
            &typesInfo,
            sizeof(typesInfo),
            NULL) == STATUS_INFO_LENGTH_MISMATCH)
        {
            if (TypeIndex < 2 || TypeIndex > typesInfo.NumberOfTypes)
            {
                return STATUS_INVALID_PARAMETER;
            }
        }
        else
        {
            return STATUS_UNSUCCESSFUL;
        }
    }
```

Reply:
That might be fine. Although I'll have to check some internals to be certain. Ideally, since we would be using some dyndata anyway - we would extract both the table and the count to avoid the syscall. But it sounds like `KphOpenObjectByTypeIndex` isn't strictly necessary for the changes anyway. I have an idea to be able to provide an implementation without dyndata at all. And might also provide an additional API... I'll get back to you on this.